### ISSUE TITLE
different interpolation between Glyphs.app and MutatorMath

### ISSUE THREAD

**Issue by marekjez86** [googlefonts/fontmake#264]
The metrics for the fonts produced by the pipeline and by the glyphs.app are NOT the same. In most of the cases I examined the fonts produced by the pipeline are slightly "taller" or slightly "fatter". While the differences are not that big, they might cause a font designer to exclaim: "Pipeline is NOT producing the font I designed" :-) 

The following was shared with me via e-mail (the fatter [green width: 803] is from pipeline and the thinner [red width: 789] is from the glyphs.app).

source used: https://github.com/googlei18n/noto-source/tree/master/src/NotoSansTamil
I'm not sure about the weight and width of the image below, but I've done my own research (please see below the following image).

![unnamed](https://cloud.githubusercontent.com/assets/13322623/23381647/111d4080-fcf4-11e6-9a5e-aaee81f6668b.png)

In my experiments I used notodiff (see https://github.com/googlei18n/nototools/tree/master/nototools ) with NotoSansTamil*.otf from https://github.com/googlei18n/noto-fonts/tree/master/alpha/from-glyphsapp/unhinted/otf/sans as --before fonts and NotoSansTamil*.otf from https://github.com/googlei18n/noto-fonts/tree/master/alpha/from-pipeline/unhinted/otf/sans as the --after font.

When generating images I set 3% (0.03) as the diff-threshhold value (i.e., ignored differences smaller than 3%).

Here are some of the difference samples (green indicates pipeline generated font, red the glyphs.app and black means common between the both fonts). I picked iitamil 0B88 for this, but there are many others:

![iitamil](https://cloud.githubusercontent.com/assets/13322623/23383912/70b30940-fcfd-11e6-98ac-171d9584cd73.png)
![iitamil 1](https://cloud.githubusercontent.com/assets/13322623/23383914/70b4055c-fcfd-11e6-91df-450267b1e175.png)
![iitamil 2](https://cloud.githubusercontent.com/assets/13322623/23383913/70b30c88-fcfd-11e6-9357-115088435e70.png)
![iitamil 3](https://cloud.githubusercontent.com/assets/13322623/23383909/70ae409a-fcfd-11e6-88bb-816f1c8cb916.png)
![iitamil 4](https://cloud.githubusercontent.com/assets/13322623/23383910/70b023c4-fcfd-11e6-8099-5ef587a98a4d.png)
![iitamil 5](https://cloud.githubusercontent.com/assets/13322623/23383911/70b0d3d2-fcfd-11e6-99c4-b12f812d92ac.png)
![iitamil 6](https://cloud.githubusercontent.com/assets/13322623/23383915/70c28816-fcfd-11e6-8eae-3cdc6ebe6809.png)
![iitamil 7](https://cloud.githubusercontent.com/assets/13322623/23383916/70c4400c-fcfd-11e6-8b25-4b5d800cee57.png)
![iitamil 8](https://cloud.githubusercontent.com/assets/13322623/23383917/70c7c3c6-fcfd-11e6-9209-f13d4bb8cd0c.png)
![iitamil 9](https://cloud.githubusercontent.com/assets/13322623/23383918/70c7f54e-fcfd-11e6-84ba-7599a75ccb9b.png)

I guess the above shape is boring :-)

Here's a difference with jaiivowelsigntamil 
NotoSansTamil-CondensedBold
![jaiivowelsigntamil](https://cloud.githubusercontent.com/assets/13322623/23384210/bded43a0-fcfe-11e6-81a9-60b0dbc09c34.png)
NotoSansTamil-ExtraCondensedBold
![jaiivowelsigntamil 1](https://cloud.githubusercontent.com/assets/13322623/23384207/bdeb12d8-fcfe-11e6-9141-fa5bc14a1e98.png)
![jaiivowelsigntamil 2](https://cloud.githubusercontent.com/assets/13322623/23384208/bdeb5694-fcfe-11e6-8254-8a8807432011.png)
![jaiivowelsigntamil 3](https://cloud.githubusercontent.com/assets/13322623/23384209/bdebf4a0-fcfe-11e6-8ce7-e1194c944969.png)
![jaiivowelsigntamil 4](https://cloud.githubusercontent.com/assets/13322623/23384212/bdf4f762-fcfe-11e6-9546-1814ec1a2310.png)
![jaiivowelsigntamil 5](https://cloud.githubusercontent.com/assets/13322623/23384211/bdf22d48-fcfe-11e6-9f4e-d9f5e2645cc1.png)
![jaiivowelsigntamil 6](https://cloud.githubusercontent.com/assets/13322623/23384213/bdfa62b0-fcfe-11e6-990d-11cecd868aea.png)
![jaiivowelsigntamil 7](https://cloud.githubusercontent.com/assets/13322623/23384214/bdfc1fb0-fcfe-11e6-9f5b-facb4c7940f6.png)
![jaiivowelsigntamil 8](https://cloud.githubusercontent.com/assets/13322623/23384216/be0045d6-fcfe-11e6-8f8b-89abb60892fa.png)
![jaiivowelsigntamil 9](https://cloud.githubusercontent.com/assets/13322623/23384215/bdfdcbda-fcfe-11e6-91c3-69487f6c43f2.png)
![jaiivowelsigntamil 10](https://cloud.githubusercontent.com/assets/13322623/23384217/be011e66-fcfe-11e6-97d1-d70f3482ae52.png)
![jaiivowelsigntamil 11](https://cloud.githubusercontent.com/assets/13322623/23384218/be03cca6-fcfe-11e6-8d04-d65376e589f8.png)

an example where height is changing:
bracketleft.tamil from NotoSansTamil-ExtraCondensedBold
![bracketleft tamil](https://cloud.githubusercontent.com/assets/13322623/23384385/8ef8f61a-fcff-11e6-87f5-ebc6b19b1396.png)

Here's the entire collection of significant differences from NotoSansTamil-ExtraCondensedBold

[NotoSansTamil-ExtraCondensedBold.zip](https://github.com/googlei18n/fontmake/files/805266/NotoSansTamil-ExtraCondensedBold.zip)

Note when I ask for the differences above 3% I get them in the following fonts:
NotoSansTamil-CondensedBold
NotoSansTamil-CondensedExtraBold
NotoSansTamil-CondensedMedium
NotoSansTamil-CondensedSemiBold

NotoSansTamil-ExtraCondensedBold
NotoSansTamil-ExtraCondensedExtraBold
NotoSansTamil-ExtraCondensedMedium
NotoSansTamil-ExtraCondensedSemiBold

NotoSansTamil-SemiCondensedBold
NotoSansTamil-SemiCondensedSemiBold

Note when I ask for the differences above 0.03% I get them in the following fonts:
NotoSansTamil-CondensedBlack
NotoSansTamil-CondensedBold
NotoSansTamil-CondensedExtraBold
NotoSansTamil-CondensedExtraLight
NotoSansTamil-CondensedMedium
NotoSansTamil-CondensedSemiBold

NotoSansTamil-ExtraCondensedBold
NotoSansTamil-ExtraCondensedExtraBold
NotoSansTamil-ExtraCondensedLight
NotoSansTamil-ExtraCondensedMedium
NotoSansTamil-ExtraCondensedSemiBold

NotoSansTamil-SemiCondensedBold
NotoSansTamil-SemiCondensedSemiBold

**Comment by marekjez86** [comment]
@behdad , @anthrotype : what do you think about the differences? is this something that we should learn to live with?

@JelleBosmaMT : FYI

**Comment by JelleBosmaMT** [comment]
This is definitely something that nobody should live with.

You will notice that the distortions are centered around the NotoSansTamil-ExtraCondensedBold. This has interpolation values weight 151 width 70 and is supposed to be a straight interpolation between the masters at 90-70 and 190-70.
What is special about weight 151-70? This design has 7 masters: 4 in the corners and two intermediates for width 100 at 90 and 151 and one intermediate for width 70 at 90.

The interpolation algorithm must have assumptions about the location of masters, that is now being tested by a real world example that doesn't match the assumption.
I only have been forwarded 2 screen shots with differences by our QA team, which is not enough to figure out what may be going wrong. But let me conclude by pointing out where we see distortions in issue #225: In the NotoSansSinhala-ExtraCondensedBold... another design with 7 masters.

**Comment by JelleBosmaMT** [comment]
With regards to the height change of the bracketleft.tamil: all masters have the same height except the narrow thin (26-70). But that should not have any effect on 151-70. Or has it?

**Comment by anthrotype** [comment]
Thank you Marek and Jelle for the thorough bug report.

I've examined the sources and the outputted fonts from both Glyphs.app and fontmake (via glyphsLib and MutatorMath), and can confirm the above mentioned differences in the interpolated instances.

As Jelle pointed out, it seems like, when we generate instances with MutatorMath, the seventh "SemiBold" master (weight=151, width=100) placed in between the "Regular" (90, 100) and "Bold" masters (190, 100) *distorts* the interpolation for the condensed instances around weight=151, e.g. "ExtraCondensed Bold" (weight=151, width=70).

On the other hand, Glyphs.app seems to not be affected by such asymmetric arrangement of the masters (4 weights for the normal width, 3 weights for the condensed width).

Here's the output from notodiff for "jatamil" and "two" glyphs in ExtraCondensed Bold, the red outline is Glyphs.app's, the green is MutatorMath's:

![notodiff_jatamil](https://dl.dropbox.com/s/2feh9skz822m2uu/A_.png?dl=1)  ![notodiff_2](https://dl.dropbox.com/s/xtooxsklr05sqza/two.png?dl=1)

I was also able to replicate this situation using the .designspace generated by glyphsLib (found inside "master_ufo" folder that fontmake produces) with Superpolator 3, which like fontmake uses MutatorMath under the hood. There one can quickly "mute" masters and see the results updated live and interactively. 

One thing I noticed is that the "ExtraCondensed Bold" (151, 70) is interpolated exactly the same way in Glyphs.app whether or not that "SemiBold" (151, 100) master is defined. That is, if I delete the "SemiBold" master, in Glyphs.app I get the same output for the "ExtraCondensed Bold" instance.

Similarly, if I delete "SemiBold" (151, 100) master and interpolate "ExtraCondensed Bold" through MutatorMath, I get the same result as Glyphs.app.

This leads me to think that if one were to add another intermediate master between "Condensed" (90, 70) and "Condensed Bold" (190, 70) masters, then the outputs from Glyphs.app and MutatorMath would agree? I still haven't tried that.. @JelleBosmaMT do you think this could be viable workaround?

In conclusion, it's not clear to me whether this is a bug in MutatorMath, or rather a "feature".

I'd be curious to know what @schriftgestalt or @LettError have to say about these different interpolations between their respective editors.
 
To help debugging, I'm attaching below a small .glyphs file containing just a couple of outlines from NotoSansTamil, that reproduces the current issue when exported from Glyphs.app and fontmake:

https://www.dropbox.com/s/a8cv56ga2p8d1zl/test_interpolation.zip?dl=0

(There's a AWS outage affecting Github so I can't upload attachments, so I'm using my own Dropbox)

**Comment by twardoch** [comment]
Here’s what I "sort of know", or at least my understanding of how design spaces differ in MutatorMath and Glyphs: 

Both **Superpolator** / **MutatorMath** and **OpenType Variations** iterate through each axis and determine the interpolated coordinates per each axis individually, step by step. This allows an arbitrary number of axes, but at a cost — predictable results are given if the masters are at "corresponding" locations, like you can see on the below image. In other words, in MutatorMath, the design space always needs to be divided into rectangular areas: 

![](http://new.superpolator.com/wp-content/uploads/2013/10/sp3_mainWindow.gif)

**Glyphs** is different: it throws all the masters from all the axes into the interpolation mix at the same time, so with each additional axis, the complexity of the equation grows by a degree. This allows greater flexibility in the arrangement of masters, but makes it much more difficult to add more axes. This is why Glyphs has the limit of 3 axes. In Glyphs, different equations are used depending on whether you have 1 axis, or 2 axes or 3. 

My understanding is that in MutatorMath, and I think also in OT Variations, your designspace always needs to be divided into rectangular areas. In other words, in Glyphs, you can have a setup like this: 

![](https://glyphsapp.com/content/2-tutorials/52-multiple-masters-part-1-setting-up-masters/interpolation-4.png)

but in MutatorMath or OT Variations, you cannot.

**Comment by twardoch** [comment]
Ps. The way I understand it, the current setup where `fontmake` just takes Glyphs masters and their locations on the design space, and then applies a completely different interpolation algorithm (MutatorMath) is "naive" and won’t ever guarantee full compatibility. I think in order to make arbitrary Glyphs design spaces work reliably in MutatorMath or OT Variations, one would need to generate more masters to build a design space that only has non-rectangular areas. 

Unfortunately, AFAIK, the Glyphs interpolation model isn’t public, so it’s hard to tell what it’s actually doing. :)

**Comment by LettError** [comment]
Hi Cosimo, 

> On 28 Feb 2017, at 22:28, Cosimo Lupo <notifications@github.com> wrote:

> In conclusion, it's not clear to me whether this is a bug in MutatorMath, or rather a "feature".
> I'd be curious to know what @schriftgestalt or @LettError have to say about these different interpolations between their respective editors.

It seems Glyphs chose a different algorithm in which certain constellations of masters are preprogrammed, but the code is proprietary and there is no point in me guessing. If Monotype delivered data that depends on such a proprietary algorithm then that is not necessarily a bug in MutatorMath.

However, if the data can be used to generate working variable fonts *using fontTools*, then I would kindly ask another look at https://github.com/fonttools/fonttools/issues/671.

> This leads me to think that if one were to add another intermediate master between "Condensed" (90, 70) and "Condensed Bold" (190, 70) masters, then the outputs from Glyphs.app and MutatorMath would agree? I still haven't tried that.. @JelleBosmaMT do you think this could be viable workaround?

You could calculate intermediates and insert them as masters, completing the rectangles. That should be relatively easy to test.

E

**Comment by JelleBosmaMT** [comment]
I have the hunch if I am going to work out in detail how this works, my boss would like to quote for it first. But I can solve it for my 7 masters.

All instances are within a rectangle. You can connect the masters by solid lines. So in my Tamil and Sinhala x = width, y is height. One intermediate master left, two on the right.

![interpol](https://cloud.githubusercontent.com/assets/12729011/23453865/4d957a7e-fe6a-11e6-9f11-aeb7f5ebeeb5.png)

All instances that are on a solid line are interpolated between the nearest masters. That is why the "missing master" at 151-70 has nothing to do with either 151-100, nor with 26-70 (or vice versa).
It also shows the relative ease to add a missing master. In fact, we can do this in under two seconds in GlyphsApp, as we only have to select the instance and select a "Add as master" button. However as we know from accessing instances trough Python, GlyphsApp handles them all as fractional values in the interpolations. Having a physical master adds rounding. So the output is slightly different and having additional rounding is bad: 1000 units per EM is crude as it is.

So for any instance that is within the rectangle, but not on a solid line, you need intermediate interpolation masters that are on a solid line. The additional challenge that is caused by the master without horizontal solid line, is to get an additional intermediate master at the same height.

**Comment by LettError** [comment]
> On 01 Mar 2017, at 10:57, Jelle Bosma <notifications@github.com> wrote:
> 
> However as we know from accessing instances trough Python, GlyphsApp handles them all as fractional values in the interpolations. Having a physical master adds rounding.

FWIW, UFOs can store floating point values for intermediates so that is not a problem.

E

**Comment by JelleBosmaMT** [comment]
On 01-mrt.-17 10:55, "Erik van Blokland" <notifications@github.com> wrote:

>It seems Glyphs chose a different algorithm in which certain
>constellations of masters are preprogrammed, but the code is proprietary
>and there is no point in me guessing. If Monotype delivered data that
>depends on such a proprietary algorithm then that is not necessarily a
>bug in MutatorMath.

Anyone working on software that has to generate font from GlyphsApp files
has a hell of a job testing all possible settings and configurations, not
just interpolation. The amount of configurations of  masters seems to be
limited by the UI, it is still a lot.

But interpolation itself does not seem to be something that depends on
math that is much more than Xi = X1 + a * (X2 - X1). The challenge is to
go through steps to decide where X1 and X2 that are needed for Xi are.

In our case, things are at their worst at the “missing master” instance
151-70. What is mysterious to me is how is is possible to end up with an
interpolation that is not using X1 and X2 from masters 90-70 and 190-70. I
can imagine algorithms that may depend on certain set configurations of
masters for the instances inside the rectangle, but in this case it seems
so bloody obvious. So a possibility is that “MutatorMath” can handle it
(to some extend), but is not given the right data.

**Comment by schriftgestalt** [comment]
Adam explained it quite accurately (only that it uses the same algorithms for all master setups and the algorithm actually supports an unlimited number of masters, Glyphs internally can use 5 (the settings are accessible by script). Only the UI is limiting it, still.

The .glyphs file contains the interpolation values (`instance.instanceInterpolations`) for each instance. So you could skip mutatormath and interpolate yourself.

**Comment by JelleBosmaMT** [comment]
Indeed it is all there. Almost too easy ;-)

**Comment by marekjez86** [comment]
@anthrotype , @behdad , @brawer : whom could make this happen?

**Comment by anthrotype** [comment]
I don't think we are planning to "skip MutatorMath" anytime soon, however "easy" this may seem to some of you.
What is relatively easier (i.e. wouldn't require changes to the code) would be to modify the .glyphs sources that have such non-rectangular interpolation space, and add intermediate masters to make it so. In theory, this could (should) be done automatically in glyphsLib, but I am a bit swamped at the moment.

**Comment by JelleBosmaMT** [comment]
On 20-mrt.-17 21:44, "Cosimo Lupo" <notifications@github.com> wrote:

>What is relatively easier (i.e. wouldn't require changes to the code)
>would be to modify the .glyphs sources that have such non-rectangular
>interpolation space, and add intermediate masters to make it so.

The task of the “pipeline” software is to generate fonts from .glyphs
sources. It is not the job of font generating software to modify  source
files and introduce an extra layer of rounding. These fonts are at a crude
1000 units per EM and there is plenty of rounding already.

And indeed, interpolating is not hard if the instance definitions already
specify masters and interpolation values. There are different ways of
doing. You can consider:


Get list of used masters with scale values needed for instance( instance )

Make a duplicate of  master( first master )

If first scale <> 1 : scale master (first duplicate, first scale)

If number of used masters > 1 then for each additional master n:
{
Make a duplicate of  master( n master )

scale master (n duplicate, n scale)

Add coordinates (n duplicate, first duplicate)
Remove master (n duplicate )
}

Round all coordinates (first duplicate)

I assume this is all done with UFO’s. There is a lot of talk how wonderful
UFO based work-flows are and how easy it all is. If half of it is true
there must be ready made code that do most of the things described above.
According to Erik van Blokland you can store fractional values so that is
not a problem either.
>

**Comment by LettError** [comment]
Here are some pictures that show how VariationModel calculates the factors, and how to interpret a system with non-aligned masters. http://letterror.com/dev/mutmath_varmod/

**Comment by JelleBosmaMT** [comment]
I just realized, that if @anthrotype can use GlyphsLib to automatically add "missing masters", then GlyphsLib can generate ANY interpolated instance. Which would solve the problem too.

**Comment by justanotherfoundry** [comment]
> This allows greater flexibility in the arrangement of masters, but makes it much more difficult to add more axes. This is why Glyphs has the limit of 3 axes.

Just a quick note, the 3-axis limitation of Glyphs is not a restriction of the algorithm that computes the coefficients. It can handle an unlimited number of axes and an unlimited number of masters.

**Comment by LettError** [comment]
@justanotherfoundry great, publish it.

**Comment by behdad** [comment]
Anyone who can reproduce the issue wants to try this patch?
```patch
diff --git a/Lib/fontTools/varLib/models.py b/Lib/fontTools/varLib/models.py
index 10825ac8..3d10921c 100644
--- a/Lib/fontTools/varLib/models.py
+++ b/Lib/fontTools/varLib/models.py
@@ -233,7 +233,10 @@ class VariationModel(object):
 				if not axis in loc:
 					continue
 				locV = loc[axis]
-				box[axis] = (self.lowerBound(locV, values), locV, self.upperBound(locV, values))
+				if locV > 0:
+					box[axis] = (0, locV, max({locV}|values))
+				else:
+					box[axis] = (min({locV}|values), locV, 0)
 
 			locAxes = set(loc.keys())
 			# Walk over previous masters now
```

**Comment by behdad** [comment]
Thanks Cosimo. Sorry about that.

I fixed varLib for, I hope this issue, and the issues that @LettError pointed out at TYPO Labs 2017. Erik, care to give it another go? It should perform as you expect now.

**Comment by behdad** [comment]
Err. This:
https://github.com/fonttools/fonttools/commit/42bef176a33dc7321eb70f5d49c90317e59ad0e5

**Comment by anthrotype** [comment]
Yay, it's working! :tada: 
Above is NotoSansTamil-VF.ttf before the patch, below is with behdad's commit in fonttools master: 
![image](https://user-images.githubusercontent.com/6939968/37964655-2b98fcd2-31ba-11e8-8fb5-557616917f38.png)

notice the bottom of the shoulder of number 2, it breaks in the top one whereas appears smooth in the bottom. Also the overall weight appears lighter for the bottom one (in fact it is slightly more compact).

This only fixes the variable font pipeline for asymmetric designspaces; for static fonts we'll need to figure out how to use Variation Model for instantiating UFOs.

**Comment by moyogo** [comment]
> This only fixes the variable font pipeline for asymmetric designspaces; for static fonts we'll need to figure out how to use Variation Model for instantiating UFOs.

Why not generate static fonts from the variable font?

**Comment by anthrotype** [comment]
well, for one thing, that currently only works for TTFs.

**Comment by anthrotype** [comment]
also, the outlines are converted interpolatably for the varfont's masters, whereas for a single instance they could be converted slightly more efficiently if taken by itself. And we don't remove overlaps for interpolatable TTF masters that make up the varfont, whereas we do for the static TTFs.

**Comment by behdad** [comment]
Yeah we should import ufoLib, add math operations to our tiny UFO objects, and run varLib on them to interpolate before creating static fonts. Basically replace mutatorMath.

**Comment by behdad** [comment]
I made another, I hope, "improvement" to the model:

  https://github.com/fonttools/fonttools/commit/a523a697164a4cdc721a06dec80791d4b1496f12

There's one more addition coming, but then that would be about it.

**Comment by marekjez86** [comment]
I like the results :-)

**Comment by madig** [comment]
Can this be closed?

**Comment by marekjez86** [comment]
Seems to be done.

**Comment by anthrotype** [comment]
@LettError Hey Eric, your http://letterror.com/dev/mutmath_varmod/ is no longer online, have you still got a copy, do you mind uploading it again somewhere? This difference b/w varLib and MutatorMath came up in discussions with a colleague and would be nice to read your handy summary. Thanks

**Comment by anthrotype** [comment]
@LettError thanks!

**Comment by belluzj** [comment]
Thanks a lot, that page is great, I was looking for it the other day and wondered where it was!
